common/flatpak-prune.c

```diff
@@ -0,0 +1,813 @@
+  g_autoptr(GHashTable) all_collection_refs = NULL;  /* (element-type OstreeChecksumRef utf8) */
+  g_autoptr(GHashTable) checksums = NULL;  /* (element-type const char *) */
+
+  checksums = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+
+  /* Get all commits up to depth from the regular refs */
+  if (!ostree_repo_list_refs (repo, NULL, &all_refs,
+                              cancellable, error))
+    return FALSE;
+
+  GLNX_HASH_TABLE_FOREACH_V (all_refs, const char*, checksum)
+    {
+      if (!traverse_commit_parents_unlocked (repo, checksum, depth, checksums, cancellable, error))
+        return FALSE;
+    }
+
+  /* Get all commits up to depth from the collection refs */
+  if (!ostree_repo_list_collection_refs (repo, NULL, &all_collection_refs,
+                                         OSTREE_REPO_LIST_REFS_EXT_EXCLUDE_REMOTES, cancellable, error))
+    return FALSE;
+
+  GLNX_HASH_TABLE_FOREACH_V (all_collection_refs, const char*, checksum)
+    {
+      if (!traverse_commit_parents_unlocked (repo, checksum, depth, checksums, cancellable, error))
+        return FALSE;
+    }
+
+  /* Find reachable objects from each commit checksum */
+  GLNX_HASH_TABLE_FOREACH_V (checksums, const char*, checksum)
+    {
+      g_autoptr(GVariant) extra_commitmeta = NULL;
+      g_autoptr(GVariant) commit_reachable = NULL;
+      FlatpakOstreeObjectName commit_name;
+
+      /* Early bail-out if we already scanned this commit in the first phase (or via some other branch) */
+      flatpak_ostree_object_name_serialize (&commit_name, checksum, OSTREE_OBJECT_TYPE_COMMIT);
+      if (object_name_bag_contains (reachable, &commit_name))
+        continue;
+
+      flatpak_debug2 ("Finding objects to keep for commit %s", checksum);
+
+      if (!load_extra_commitmeta (repo, checksum, &extra_commitmeta, cancellable, error))
+        return FALSE;
+
+      if (extra_commitmeta)
+        commit_reachable = g_variant_lookup_value (extra_commitmeta, "xa.reachable", G_VARIANT_TYPE ("a" FLATPAK_OSTREE_OBJECT_NAME_ELEMENT_TYPE));
+
+      if (commit_reachable == NULL)
+        {
+          g_autoptr(GHashTable) commit_reachable_ht = reachable_commits_new ();
+          g_autoptr(GVariant) new_extra_commitmeta = NULL;
+          g_autofree FlatpakOstreeObjectName *commit_reachable_raw = NULL;
+          FlatpakOstreeObjectName *next_commit_reachable_raw;
+          g_auto(GVariantDict) extra_commitmeta_builder = FLATPAK_VARIANT_BUILDER_INITIALIZER;
+
+          if (!ostree_repo_traverse_commit_union (repo, checksum, 0, commit_reachable_ht,
+                                                  cancellable, error))
+            return FALSE;
+
+          commit_reachable_raw = g_new (FlatpakOstreeObjectName, g_hash_table_size (commit_reachable_ht));
+
+          next_commit_reachable_raw = &commit_reachable_raw[0];
+          GLNX_HASH_TABLE_FOREACH_V (commit_reachable_ht, GVariant *, reachable_commit)
```

Comment:
Aren't these objects not commits, so should the variable be called `reachable_object`?

Reply:
Yes they are, however I was thinking of the name as meaning "reachable from this commit, as opposed to the global reachable set. Might still be some better naming, but there are a lot of similar names flying around, so its not easy to pick new ones.

Reply:
`commit_reachable_object` maybe

Reply:
Not sure, its kinda long.